state/protocol/protocol_state/epoch_fallback_statemachine.go

```diff
@@ -30,6 +30,13 @@ func newEpochFallbackStateMachine(view uint64, parentState *flow.RichProtocolSta
 	}
 }
 
+// NewEpochFallbackStateMachine instantiates a new `protocolStateMachine` and returns is as an interface type `ProtocolStateMachine`.
+// Thereby, `NewStateMachine` is a specific version of `StateMachineFactoryMethod` for creating a happy-path state machine.
+// This function never errors.
+func NewEpochFallbackStateMachine(candidateView uint64, parentState *flow.RichProtocolStateEntry) (ProtocolStateMachine, error) {
```

Comment:
is this only for sake of simplifying code in factory? I've specifically made it private so it can be created only from factory which knows about specific implementation details(what state machine to create)

Reply:
We have defined this signature for the factory method:
https://github.com/onflow/flow-go/blob/92dfe41cedb1b2ed3a6ada345b625e1417ca9552/state/protocol/protocol_state/mutator.go#L16

unfortunately, the go compiler isn't smart enough to accept the concretely-typed constructor:
https://github.com/onflow/flow-go/blob/92dfe41cedb1b2ed3a6ada345b625e1417ca9552/state/protocol/protocol_state/protocol_statemachine.go#L33
* On the one hand, we would like to have a constructor for the concretely typed structs for testing purposes (our tests inspect internal fields of the struct for correctness, which is fine in my opinion but a purist my object).
* On the other hand, the mutator's constructor needs the interface types https://github.com/onflow/flow-go/blob/92dfe41cedb1b2ed3a6ada345b625e1417ca9552/state/protocol/protocol_state/mutator.go#L53-L54    

Simplified example for experimentation: https://go.dev/play/p/NbiZCzg0Qj3?v=goprev

You had solved this via functors, which is probably the best solution if we prefer not to export the constructors (which it seems we don't want to). I will revert to the  functors